File: app/src/main/java/email/schaal/ocreader/util/FeedColors.java
package email.schaal.ocreader.util;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.palette.graphics.Palette;
import androidx.palette.graphics.Target;

import java.util.Collections;

/**
 * Generate text and background color for Feeds
 */
public class FeedColors {
    @Nullable
    private final Palette palette;

    public enum Type {
        TEXT,
        BACKGROUND
    }

    FeedColors(@NonNull Palette palette) {
        this.palette = palette;
    }

    FeedColors(@Nullable @ColorInt Integer color) {
        final Palette.Swatch swatch;

        if(color != null) {
            swatch = new Palette.Swatch(color, 1);
        } else {
            swatch = null;
        }

        if(swatch != null)
            palette = new Palette.Builder(Collections.singletonList(swatch)).addTarget(Target.MUTED).generate();
        else
            palette = null;
    }

    @ColorInt public int getColor(@NonNull Type type, @ColorInt int defaultColor) {
        final Palette.Swatch swatch;

        if(palette != null) {
            switch (type) {
                case TEXT:
                    swatch = palette.getDominantSwatch();
                    break;
                case BACKGROUND:
                    swatch = palette.getMutedSwatch();
                    break;
                default:
                    swatch = null;
                    break;
            }
        } else {
            swatch = null;
        }

        if(swatch != null)
            return swatch.getRgb();
        else
            return defaultColor;
    }
}

File: app/src/main/java/email/schaal/ocreader/util/FeedColors.kt
package email.schaal.ocreader.util

import androidx.annotation.ColorInt
import androidx.palette.graphics.Palette
import androidx.palette.graphics.Palette.Swatch
import androidx.palette.graphics.Target

/**
 * Generate text and background color for Feeds
 */
class FeedColors(private val palette: Palette?) {
    enum class Type {
        TEXT, BACKGROUND
    }

    internal constructor(@ColorInt color: Int?) : this(
            color?.let { Swatch(it, 1) }
                    ?.let { Palette.Builder(listOf(it)).addTarget(Target.MUTED).generate() })

    @ColorInt
    fun getColor(type: Type, @ColorInt defaultColor: Int): Int {
        return palette?.let {
            when (type) {
                Type.TEXT -> it.dominantSwatch
                Type.BACKGROUND -> it.mutedSwatch
            }
        }?.rgb ?: defaultColor
    }
}